### ISSUE TITLE
integrate aflgo? 

### ISSUE THREAD

**Issue by vanhauser-thc** [vanhauser-thc/AFLplusplus#20]
@strongcourage had the idea of integrating https://github.com/aflgo/aflgo into afl++

My opinion - just putting down my thoughts:

its an interesting feature for sure.

we partially can achieve something like that with the whitelist feature in llvm_mode, and @hexcoder- has thought about rewriting the whitelist feature code to allow to match on function names.
But it does not reach the fine-grained targeting that aflgo has.

usually if you fix something you have a crashing test case, only if you dont have one, you would need aflgo.

the overall steps to finally run aflgo are quite some and its complicated to get there (compared to normal llvm_mode compiling)

looking at the specific changes: https://github.com/mirrorer/afl/compare/master...aflgo:master
I can see that the shm mapsize is different to normal, and the llvm_mode has been rewritten a lot.
So: I don't think the two projects are actually compatible to have in one tool - or only with complex code changes

TLDR: interesting feature, however very special case, and I doubt it can be easily integrated.  However I only had a quick look

**Comment by strongcourage** [comment]
I agree that the most complex part comes from llvm_mode including prepocessing mode and distance instrumentation mode. Also, we need to use clang v4.0 (I have not tested newer versions) and LLVMgold plugin.

However, I had a forked repo with [build & fuzz](https://github.com/strongcourage/aflgo/tree/master/scripts) scripts.

**Comment by vanhauser-thc** [comment]
@strongcourage llvm 4 is not on modern Linux distributions anymore so this would need to be updated.
and to make sense it would need to work with llvm 4 to 8 (like afl++: llvm_mode supports llvm 3.8 to 8).

@hexcoder- has now some experience with the gold plugin, but as far as I have understood him he is not happy to have gained that experience ;)

I am not convinced that a) its feasible to integrate, and b) that it makes the project better. because we will need to support llvm 9 later this year, and the more llvm code we have to maintain and migrate, the harder the burden.

@strongcourage how would you imagine the aflgo feature being implemented? specifiying an AFL_LLVM_AFLGO=1 AFL_LLVM_AFLGO_TARGET=foo.txt when doing llvm_mode?

**Comment by strongcourage** [comment]
I think the most important thing is to try AFLGo with LLVM 8 as you suggested, since there were quite a lot of [issues](https://github.com/aflgo/aflgo/issues) of AFLGo relevant to the LLVMgold plugin and the version of Clang.

Other parts, e.g, afl-fuzz.c are quite straightforward.

**Comment by vanhauser-thc** [comment]
yes I made a diff to see the changes in afl-fuzz, and its managable.

do you think it would make sense to add it to llvm_mode or better to have an aflgo_mode/ directory for it? both have advantages and disadvantages ... in llvm_mode it makes it more complex - how would you select normal mode and aflgo? - in aflgo_mode/ it would loose the features like instrim, laf-intel etc.

dunno

**Comment by strongcourage** [comment]
The main difference comes from afl-llvm-pass.so.cc. I think it is doable to just modify this file to support both modes.

**Comment by vanhauser-thc** [comment]
well I'd say fork off afl++ and try to get it into llvm_mode then. @hexcoder- or do you have a different opinion?

as @hexcoder- knows more about llvm_mode and maintaining and gold plugin I would leave the review, feedback etc to him

**Comment by hexcoder-** [comment]
I would 
1. get aflgo, 
2. get through the issues, 
3. try to integrate proposed fixes, 
4. then make a diff against afl 2.49
5. and apply this to a new branch in llvm_mode
6. then add llvm compatibility fixes (V.8.0.0).

But unfortunately I don't have as much time as @vanhauser-thc at my disposal. But I can do it.

**Comment by strongcourage** [comment]
@hexcoder-: I can do it and we can discuss if there are problems. OK?

**Comment by hexcoder-** [comment]
Yep, thats ok with me.

**Comment by vanhauser-thc** [comment]
just curious - is there currently something happening, is it on hold (vacation etc) or not important anymore?

**Comment by strongcourage** [comment]
sorry, I am currently busy due to research deadline. get back very soon.

**Comment by vanhauser-thc** [comment]
all good, no pressure :) just wanted to know if this will progress at some point or not. take the time you need.

**Comment by vanhauser-thc** [comment]
is there anything still happening here? otherwise I will close the issue.

**Comment by andreafioraldi** [comment]
close it

**Comment by vanhauser-thc** [comment]
@strongcourage if you are ready with an implementation put a PR or reopen the issue
